src/ClientFactory.php

```diff
@@ -0,0 +1,64 @@
+<?php
+
+namespace Madmatt\SilverStripeS3;
+
+use LogicException;
+
+abstract class ClientFactory
+{
+    /**
+     * @return string
+     */
+    protected function findAwsKeyId()
+    {
+        if (getenv('AWS_ACCESS_KEY_ID') !== false) {
+            return (string) getenv('AWS_ACCESS_KEY_ID');
+        }
+
+        return '';
+    }
+
+    /**
+     * @return string
+     */
+    protected function findAwsSecretKey()
+    {
+        if (getenv('AWS_SECRET_ACCESS_KEY') !== false) {
+            return (string) getenv('AWS_SECRET_ACCESS_KEY');
+        }
+
+        return '';
+    }
+
+    /**
+     * @return string
+     */
+    protected function findAwsRegion()
+    {
+        if (getenv('AWS_REGION') !== false) {
+            return (string) getenv('AWS_REGION');
+        }
+        throw new LogicException('AWS_REGION environment variable not set');
+    }
+
+    /**
+     * @return array
+     */
+    protected function getConnectionArgs()
+    {
+        $args = [
+            'region' => $this->findAwsRegion(),
+            'version' => 'latest',
+        ];
+        $key = $this->findAwsKeyId();
+        $secret = $this->findAwsSecretKey();
+        if ($key && $secret) {
+            $args['credentials'] = [
+                'key' => $key,
```

Comment:
Passing the key and secret this way is I think redundant. `ClientResolver` that runs on AwsClient constructor should awtomatically get creds from the environment, which has an added advantage of working with EC2 roles. The only one that is required is probably the region.